lib/tree.rb

```diff
@@ -16,40 +16,52 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n) - number of stack frames relative to the input
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    return @root = new_node if @root.nil?
+    add_helper(@root, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n) - number of stack frames relative to the input
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) - number of stack frames relative to the input + result array
   def inorder
-    raise NotImplementedError
+    result = []
+    return result if @root.nil?
+    inorder_helper(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) - number of stack frames relative to the input + result array
   def preorder
-    raise NotImplementedError
+    result = []
+    return result if @root.nil?
+    pre_order_helper(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) - number of stack frames relative to the input + result array
   def postorder
-    raise NotImplementedError
+    result = []
+    return result if @root.nil?
+    post_order_helper(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n) - number of stack frames relative to the input
   def height
```

Comment:
👍 , the space complexity would be O(log n) if the tree is balanced and O(n) otherwise.

Reply:
But even with the balanced tree we will need to visit every node? That's why it looks like O(n) to me.